### PR TITLE
Panorama c02

### PR DESCRIPTION
**Pull request by RottenKiwi**
Final pull request for Panorama Swap audit C02 issue.

### CODE REVIEW

File: trading_pair_azero/lib.rs [hunk 1]
```diff
@@ -55,98 +60,195 @@ pub mod trading_pair_azero {
...
+            let contract_overall_generated_psp22_fee: Balance = 0; // Total PSP22 fees generated by the contract
+            let contract_overall_generated_azero_fee: Balance = 0; // Total AZERO fees generated by the contract
+            let psp22_to_give_in_a_day = Mapping::default(); // Mapping to store daily PSP22 fees to be given
+            let account_overall_staking_rewards = Mapping::default(); // Mapping to store overall staking rewards for accounts
+            let account_overall_lp_fee_rewards = Mapping::default(); // Mapping to store overall LP fee rewards for accounts
+            let last_redeemed = Mapping::default(); // Mapping to store last redeemed time for accounts
+            let staking_percentage = 2; // Percentage of fees to be distributed as staking rewards
+            let tpa_contract_address: AccountId = Self::env().account_id();
+
+            psp22
```
Comment (coreggon11) [inline trading_pair_azero/lib.rs:246]: this is not needed

File: trading_pair_azero/lib.rs [hunk 2]
```diff
@@ -55,98 +60,195 @@ pub mod trading_pair_azero {
...
+        caller: AccountId, // Address of the caller who initiated the PSP22 token swap
+        psp22_deposited_amount: Balance, /* Amount of PSP22 tokens deposited by the caller for the swap */
+        a0_given_amount: Balance, // Amount of AZERO tokens given to the caller as part of the swap
+        a0_given_to_vault: Balance, // Amount of AZERO tokens sent to the vault as part of the swap
     }
 
+    impl PSP22 for TradingPairAzero {
+        #[ink(message)]
+        fn allowance(&self, owner: AccountId, spender: AccountId) -> Balance {
+            self.lp_tokens_allowances
```
Comment (coreggon11) [inline trading_pair_azero/lib.rs:125]: this field is also in psp22::Data, so if you are using the OB implementation, there is no need for this field, same for the balances field. Also, the implementation seems to be the same the one from OB, so another question is if you need this custom implementation (it is okay if it is kept, just a question to ask)

Comment (RottenKiwi) [reply trading_pair_azero/lib.rs:125]: We wish to use both allowances and both balances, yes :)

File: trading_pair_azero/lib.rs (hunk 1 above)
Comment (RottenKiwi) [reply trading_pair_azero/lib.rs]: Removed and fixed.

**Comment by RottenKiwi** [comment]
Took care of all the comments, merging with main.

**Pull request merged**